File: C++/src/raytrace-cpp-lib/ray.cpp
/**
 *  This file is a part of raytrace-cpp.
 *  Copyright (C) 2023-2024  Amani Hernandez
 *
 *  raytrace-cpp is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU Affero General Public License as published
 *  by the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU Affero General Public License for more details.
 *
 *  You should have received a copy of the GNU Affero General Public License
 *  along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

#include <raytrace-cpp-lib/vec3.hpp>
#include <raytrace-cpp-lib/ray.hpp>

ray::ray() {}

ray::ray(const vec3<float>& origin, const vec3<float>& direction)
    : origin(origin), direction(direction) {}

vec3<float> ray::get_origin() const
{
    return origin;
}

vec3<float> ray::get_direction() const
{
    return direction;
}

void ray::set_direction(const vec3<float>& direction)
{
    this->direction = direction;
}

vec3<float> ray::point_at(float t) const
{
    return origin + direction * t;
}

vec3<float> ray::lerp(float t) const
{
    return origin * (1.0 - t) + direction * t;
}

File: Java/src/main/java/kz/sn34/raytrace-java-lib/Ray.java
package kz.sn34.raytrace_java_lib;

public class Ray
{
    private Vector3 origin;
    private Vector3 direction;

    public Ray()
    {
        this.origin = new Vector3();
        this.direction = new Vector3();
    }

    public Ray(Vector3 origin, Vector3 direction)
    {
        this.origin = origin;
        this.direction = direction;
    }

    public Vector3 getOrigin()
    {
        return this.origin;
    }

    public Vector3 getDirection()
    {
        return this.direction;
    }

    public void setOrigin(Vector3 origin)
    {
        this.origin = origin;
    }

    public void setDirection(Vector3 direction)
    {
        this.direction = direction;
    }

    public Vector3 pointAt(double t)
    {
        return this.origin.add(this.direction.multiply(t));
    }

    public Vector3 lerp(double t)
    {
        return this.origin.multiply(1 - t).add(this.direction.multiply(t));
    }
}
